krm-functions/configinject-fn/fn/function.go

```diff
@@ -139,84 +141,120 @@ func (f *FnR) desiredOwnedResourceList(forObj *fn.KubeObject) (fn.KubeObjects, e
-			kfko := rl.Items.GetRootKptfile()
-			if kfko == nil {
-				return nil, fmt.Errorf("mandatory Kptfile is missing from the package %s, repo %s", pr.Spec.PackageName, pr.Spec.RepositoryName)
-			}
-			kf := kptfilelibv1.KptFile{Kptfile: kfko}
 
-			// get the dependency objects in the package and check its status
-			gvk := schema.GroupVersionKind{
-				Group:   nephiodeployv1alpha1.GroupVersion.Group,
-				Version: nephiodeployv1alpha1.GroupVersion.Version,
-				Kind:    nephiodeployv1alpha1.UPFDeploymentKind,
-			}
-			for _, ref := range dep.Spec.Injectors {
-				/*
-					for _, o := range rl.Items {
-						fn.Logf("configinject resource apiVersion: %s kind: %s\n", o.GetAPIVersion(), o.GetKind())
-					}
-				*/
-				fn.Logf("configinject dependency gvk: %s\n", gvk.String())
+			prName := fmt.Sprintf("%s-%s", pr.Spec.RepositoryName, pr.Spec.PackageName)
 
-				depObjs := rl.Items.Where(fn.IsGroupVersionKind(ref.GroupVersionKind()))
-				if len(depObjs) == 0 {
-					fn.Logf("configinject dependency not ready: the package %s in repo %s does not contain a resource with %s\n", pr.Spec.PackageName, pr.Spec.RepositoryName, gvk.String())
-					return nil, fmt.Errorf("dependency not ready: the package %s in repo %s does not contain a resource with %s", pr.Spec.PackageName, pr.Spec.RepositoryName, gvk.String())
-				}
-				for _, o := range depObjs {
-					ct := kptfilelibv1.GetConditionType(&corev1.ObjectReference{
-						APIVersion: o.GetAPIVersion(),
-						Kind:       o.GetKind(),
-						Name:       o.GetName(),
-					})
-					c := kf.GetCondition(ct)
-					if c == nil {
-						fn.Logf("configinject dependency not ready: the package %s in repo %s does not contain a condition for %s\n", pr.Spec.PackageName, pr.Spec.RepositoryName, ct)
-						return nil, fmt.Errorf("dependency not ready: the package %s in repo %s does not contain a condition for %s", pr.Spec.PackageName, pr.Spec.RepositoryName, ct)
-					}
-					if c.Status != kptv1.ConditionTrue {
-						// we fail fast if the condition is not true
-						fn.Logf("configinject dependency not ready: the package %s in repo %s has a condition which is False for: %s\n", pr.Spec.PackageName, pr.Spec.RepositoryName, c.Type)
-						return nil, fmt.Errorf("dependency not ready: the package %s in repo %s has a condition which is False for: %s", pr.Spec.PackageName, pr.Spec.RepositoryName, c.Type)
+			if porchv1alpha1.LifecycleIsPublished(pr.Spec.Lifecycle) {
+				if latestPR, ok := prmap[prName]; ok {
+					// both the latest pr and the new pr are published
+					// update the map with the latest pr
+					// if the revision of the new pr is better than the one of the latest pr in the map
+					newRev, err := getRevisionNbr(pr.Spec.Revision)
+					if err != nil {
+						return nil, err
 					}
-					// encapsulates the resource in another CR
-					newObj, err := GetConfigKubeObject(forObj, o)
+					latestRev, err := getRevisionNbr(latestPR.Spec.Revision)
```

Comment:
Is this going to blow up for the "main" revision? Or for Drafts? should we map those to revision number 0 or something?

Reply:
Actually I just verified, the latest revision will have an annotation:

```
  labels:
    kpt.dev/latest-revision: "true"
```

Of course...since updates to multiple objects are non-atomic, maybe relying on your comparison is better (ie, the API server cannot guarantee that exactly one package revision has that annotation at any given time).